### ISSUE TITLE
Support spec injection when testing with a host component.

### ISSUE THREAD

**Issue by pcafstockf** [ngneat/spectator#443]
## I'm submitting a...
<pre><code>
[ ] Regression (a behavior that used to work and stopped working in a new release)
[ ] Bug report  
[ ] Performance issue
[x] Feature request
[ ] Documentation issue or request
[ ] Support request
[ ] Other... Please describe:
</code></pre>

## Current behavior
Currently spec injection works great for component testing:
```

    const spectatorFactory = createComponentFactory<FooComponent>({
        component: FooComponent
    });
    it('should render.', inject([MockGenerationService], (dataSvc: MockGenerationService) => {
        const spectator = spectatorFactory({
            props: {model: dataSvc.makeModel()}
        });
        ... expectations
    }));
```
However, spec injection does not work for host component testing because `spectatorHostFactory` tries to modify the `TestBed` after spec injection has sealed it:
```

    const spectatorHostFactory = createHostFactory<FooComponent>({
        component: FooComponent
    });
    it('should render.', inject([MockGenerationService], (dataSvc: MockGenerationService) => {
        const spectatorHost = spectatorHostFactory('<app-foo model="model"></app-foo>',  {
            hostProps: {model: dataSvc.makeModel()};
        });
    // The above throws an error about overrideModule being called after inject.
        ... expectations
    }));
```

## Expected behavior
It would be great if spec injection could be done with both component and host component.  
Something like:
```

    const spectatorHostFactory = createHostFactory<FooComponent>({
        component: FooComponent,
        template: '<app-foo model="model"></app-foo>'
    });
    it('should render.', inject([MockGenerationService], (dataSvc: MockGenerationService) => {
        const spectatorHost = spectatorHostFactory({
            hostProps: {model: dataSvc.makeModel()};
        });
        ... expectations
    }));
```

## Minimal reproduction of the problem with instructions
If this feature is approved, I would be happy to provide a pull request.  
I have it working locally with minor refactoring to the createHostFactory function while preserving existing functionality.

## What is the motivation / use case for changing the behavior?
I defer to the Angular team for an explanation as to why spec injection can be extremely useful :-)

## Environment

<pre><code>
Angular version: 11.2.x

Browser:
- [x] Chrome (desktop) version XX
- [ ] Chrome (Android) version XX
- [ ] Chrome (iOS) version XX
- [ ] Firefox version XX
- [x] Safari (desktop) version XX
- [ ] Safari (iOS) version XX
- [ ] IE version XX
- [ ] Edge version XX
 * Applicable to all, but my local fix only tested in Chrome and Safari.
 
For Tooling issues:
- Node version: NA
- Platform:  Mac

Others:
NA
</code></pre>

**Comment by NetanelBasal** [comment]
Can you explain the benefits of using it over the existing API we provide?

**Comment by pcafstockf** [comment]
> Can you explain the benefits of using it over the existing API we provide?

Busted! :grin:

Our real need is to be able to use DI before the component (or host) is created by `spectatorHostFactory`.   
Spec injection (which already works for components) is something everyone is familiar with, so rather than writing a big long winded explanation or our own needs, it just seemed simpler to request a commonly used feature (which would simultaneously address our needs).

We create a fresh component for each spec/test, but the way the component is actually created is driven to some extent by the data that will be displayed.

Our mock data is extremely complex and must be generated at runtime in order to be realistic in any way.  This mocked data interface relies on a number of other mocked services, so dependency injection is the cleanest way to scaffold all those mocks.

Bottom line is we need access to the injection framework before we ask the `spectatoFactory` or `spectatorHostFactory` to create the component.  
Currently, accessing injection locks us out of the additional configuration of TestBed that `spectatorHostFactory ` currently performs.  
We understand that's a limitation of TestBed, but this gives us a way to work around that limitation.

**Comment by NetanelBasal** [comment]
Alright, I'm willing to see a PR as long as it doesn't break anything :)
